src/svc/layer.rs

```diff
@@ -0,0 +1,49 @@
+use std::marker::PhantomData;
+
+/// A stackable element.
+///
+/// Given a `Next`-typed inner value, produces a `Bound`-typed value.
+/// This is especially useful for composable types like `MakeClient`s.
+pub trait Layer<Next> {
+    type Bound;
+
+    /// Produce a `Bound` value from a `Next` value.
+    fn bind(&self, next: Next) -> Self::Bound;
+
+    /// Compose this `Layer` with another.
+    fn and_then<M>(self, inner: M) -> AndThen<Next, Self, M>
+    where
+        Self: Layer<M::Bound> + Sized,
+        M: Layer<Next>,
+    {
+        AndThen {
+            outer: self,
+            inner,
+            _p: PhantomData,
+        }
+    }
+}
+
+/// Combines two `Layers` into one layer.
+#[derive(Debug, Clone)]
+pub struct AndThen<Next, Outer, Inner>
+where
+    Outer: Layer<Inner::Bound>,
+    Inner: Layer<Next>,
+{
+    outer: Outer,
+    inner: Inner,
+    _p: PhantomData<Next>,
```

Comment:
TIOLI: perhaps this should be `PhantomData<fn() -> Next>` or something, so that `AndThen` being `Send` or `Sync` isn't dependent on `Next` being `Send` or `Sync`?